palette/src/convert.rs

```diff
@@ -499,6 +502,173 @@ where
     }
 }
 
+///The error type for a color conversion that converted a color into a color with invalid values.
+#[derive(Debug)]
+pub struct OutOfBounds<T> {
+    color: T,
+}
+
+impl<T> OutOfBounds<T> {
+    ///Create a new error wrapping a color
+    fn new(color: T) -> Self {
+        OutOfBounds { color }
+    }
+
+    ///Consume this error and return the wrapped color
+    pub fn color(self) -> T {
+        self.color
+    }
+}
+
+impl<T: Debug> error::Error for OutOfBounds<T> {
+    fn description(&self) -> &str {
+        "Color conversion is out of bounds"
+    }
+}
+
+impl<T> Display for OutOfBounds<T> {
+    fn fmt(&self, fmt: &mut Formatter) -> fmt::Result {
+        write!(fmt, "Color conversion is out of bounds")
+    }
+}
+
+///An extension for `Result` that offers a method to take out the color of `try_convert` result.
+pub trait ResultExt<T> {
+    ///Take the color out of this `Result`.
+    fn take(self) -> T;
+}
+
+impl<T> ResultExt<T> for Result<T, OutOfBounds<T>> {
+    #[inline]
+    fn take(self) -> T {
+        match self {
+            Ok(color) => color,
+            Err(color) => color.color(),
+        }
+    }
+}
+
+///A trait for converting from one color to another.
+///
+///`convert_unclamped` currently wraps the underlying `Into` implementation.
+pub trait Convert<T>: Sized + Into<T> {
+
+    ///Convert into T with values clamped to the color defined bounds
+    ///
+    ///```
+    ///use palette::Convert;
+    ///use palette::Limited;
+    ///use palette::{Srgb, Lch};
+    ///
+    ///
+    ///let rgb: Srgb = Lch::new(50.0, 100.0, -175.0).convert();
+    ///assert!(rgb.is_valid());
+    ///```
+    fn convert(self) -> T;
+
+    ///Convert into T, the resulting color might be invalid in it's color space
+    ///
+    ///```
+    ///use palette::Convert;
+    ///use palette::Limited;
+    ///use palette::{Srgb, Lch};
+    ///
+    ///let rgb: Srgb = Lch::new(50.0, 100.0, -175.0).convert_unclamped();
+    ///assert!(!rgb.is_valid());
+    ///```
+    #[inline]
+    fn convert_unclamped(self) -> T {
+        self.into()
+    }
+
+    ///Convert into T, returning ok if the color is inside of its defined range,
+    ///otherwise an `OutOfBounds` error is returned which contains the unclamped color.
+    ///
+    ///```
+    ///use palette::Convert;
+    ///use palette::{Srgb, Hsl};
+    ///
+    ///let rgb: Srgb = match Hsl::new(150.0, 1.0, 1.1).try_convert() {
+    ///    Ok(color) => color,
+    ///    Err(err) => {
+    ///        println!("Color is out of bounds");
+    ///        err.color()
+    ///    },
+    ///};
+    ///```
+    fn try_convert(self) -> Result<T, OutOfBounds<T>>;
+}
+
+macro_rules! expand_impl_convert {
```

Comment:
I would honestly like to move away from macros if I can. They have been a hassle and I have tried to replace some of them with derives and just regular `impl`s. Do you think all of this would be possible to replace with:

```rust
impl<T: Into<U> + Limited, U> Convert<U> for T {
    // ...
}
```

It has the same pattern as the `Into` blanket implementation, so I think it should work. It's not a big deal if it's no perfect yet, and I'm sure it will have to change anyway when the rest of the library catches up and the implementation of `Into` changes. I just want a start that is somewhat in line with my longer term goals (which I maybe should note down in an issue instead of keeping in my head).

Reply:
Understandable, Macro implementations make code quite difficult to read. I was trying to get rid of the `Limited` trait-bound but that just wasnt possible with a generic `impl`. I can revert back to `impl`ementing the trait. While we are talking about the similarities to `Into`, wouldn't it make more sense to make this two traits which are `ConvertFrom` and `ConvertInto` then?